### ISSUE TITLE
Improve startup performance by removing unnecessary `executable()` checks

### ISSUE THREAD

**Issue by kyoh86** [neovim/neovim#31506]
### Problem

Neovim's `vim._defaults.lua` executes a call to `executable()` to check for the presence of `rg` and sets it as `'grepprg'`. When `rg` is not found, this process is slow and introduces unnecessary overhead during startup. Users do not expect such a heavy check in the default configuration.

This `executable()` check can cause a delay of 50–200 milli seconds, further exacerbating the performance issue. 

### Steps to reproduce

Uninstall ripgrep.
Install Neovim.
Start Neovim with nvim --clean --startuptime foo.txt.

Observe the execution delay caused by the executable check in `vim._defaults.lua`.
For example, try to delete below lines.

```
--- Default options
do
  --- Default 'grepprg' to ripgrep if available.
  if vim.fn.executable('rg') == 1 then
    -- Use -uu to make ripgrep not check ignore files/skip dot-files
    vim.o.grepprg = 'rg --vimgrep -uu '
    vim.o.grepformat = '%f:%l:%c:%m'
  end
end
```

### Expected behavior

Neovim should avoid calling executable to check for the presence of `rg` during startup. When `rg` is not found, the `executable()` check triggers an exhaustive path search, which significantly increases the startup time. Removing this check entirely would lead to improved performance, as users can configure `'grepprg'` manually if needed.

The performance cost of `executable()` is a well-known problem, and calling it during startup should be avoided unless absolutely necessary.

### Nvim version (nvim -v)

v0.11.0-dev-b1c907f

### Vim (not Nvim) behaves the same?

no

### Operating system/version

Ubuntu 24.04.1 LTS

### Terminal name/version

WezTerm

### $TERM environment variable

xterm-256color

### Installation

build from repo

**Comment by zeertzjq** [comment]
cc @LunarLambda

**Comment by LunarLambda** [comment]
I'm surprised to hear that an executable check can take 50 to 200 ms unless something is severely wrong with either the user's environment or the implementation of the function. In my testing it didn't make any meaningful dent in startup time in either the positive or negative case, and during the review process of the original PR nothing suggested there being any issue with the check.

Can the check be made asynchronous or deferred somehow so it doesn't block startup? I could put up a fix in that case, it would be unfortunate to have to remove the feature entirely.

**Comment by Shougo** [comment]
Unfortunately, `executable()` is extremely slow in Windows environment or WSL environment.
(I think it is from Security check process issue)
I think `grepprg` should be set manually by users.

> In my testing it didn't make any meaningful dent in startup time in either the positive or negative case, and during the review process of the original PR nothing suggested there being any issue with the check.

Because many users and developers uses Linux environment or very faster computer.

> Can the check be made asynchronous or deferred somehow so it doesn't block startup? I could put up a fix in that case, it would be unfortunate to have to remove the feature entirely.

I think Neovim should not set it by async.  Because it may overwrite users set value.

**Comment by kyoh86** [comment]
@LunarLambda Those who have `ripgrep` can configure it as their `'grepprg'`, and that should be sufficient.
The `'grepprg'` setting itself exists precisely for this purpose. I don't see much value in setting a completely third-party, independent external tool—one that has no inherent connection to Neovim— **as the default** with high performance cost.
Even if this feature is removed, we can still configure `ripgrep` as `'grepprg'` and use it just as we always have.

**Comment by zeertzjq** [comment]
> Can the check be made asynchronous or deferred somehow so it doesn't block startup?

`vim.fn` is only available on the main thread, so deferring the check will still block Nvim.

**Comment by LunarLambda** [comment]
I can change the check to look in a smaller set of places, and/or gate it behind a `g:` variable or similar (treating it like/turning it into a tiny built-in plugin, assuming source order allows it)

Otherwise, I don't believe my input or approval is needed for the maintainers to remove the feature should they so wish.

**Comment by zeertzjq** [comment]
`_defaults.lua` is sourced before user config

**Comment by lewis6991** [comment]
Any solution should probably only be specific for Windows. `vim.fn.executable` is fine on even slow Linux/MacOS hosts.

**Comment by LunarLambda** [comment]
Is it Windows as a whole or just WSL specifically? Do we have any means of detecting WSL consistently?

**Comment by SebasF1349** [comment]
I didn't even notice, but it's similar to me.

This is the output of `nvim --clean --startuptime log` with ripgrep installed:

```
--- Startup times for process: Primary (or UI client) ---

times in msec
 clock   self+sourced   self:  sourced script
 clock   elapsed:              other lines

000.000  000.000: --- NVIM STARTING ---
000.100  000.100: event init
000.167  000.067: early init
000.296  000.129: locale set
000.330  000.034: init first window
000.597  000.267: inits 1
000.602  000.005: window checked
000.606  000.003: parsing arguments
000.908  000.026  000.026: require('vim.shared')
000.974  000.024  000.024: require('vim.inspect')
001.002  000.023  000.023: require('vim._options')
001.003  000.093  000.046: require('vim._editor')
001.004  000.139  000.019: require('vim._init_packages')
001.006  000.261: init lua interpreter
003.986  002.980: nvim_ui_attach
004.217  000.232: nvim_set_client_info
004.218  000.001: --- NVIM STARTED ---

--- Startup times for process: Embedded ---

times in msec
 clock   self+sourced   self:  sourced script
 clock   elapsed:              other lines

000.000  000.000: --- NVIM STARTING ---
000.063  000.063: event init
000.112  000.049: early init
000.266  000.154: locale set
000.293  000.027: init first window
000.512  000.218: inits 1
000.520  000.008: window checked
000.523  000.003: parsing arguments
000.804  000.025  000.025: require('vim.shared')
000.862  000.024  000.024: require('vim.inspect')
000.891  000.025  000.025: require('vim._options')
000.893  000.087  000.039: require('vim._editor')
000.894  000.129  000.017: require('vim._init_packages')
000.895  000.244: init lua interpreter
000.924  000.029: expanding arguments
000.936  000.012: inits 2
001.116  000.181: init highlight
001.117  000.001: waiting for UI
002.290  001.173: done waiting for UI
002.299  000.009: clear screen
002.397  000.009  000.009: require('vim.keymap')
002.933  000.632  000.623: require('vim._defaults')
002.935  000.004: init default mappings & autocommands
004.427  000.775  000.775: sourcing /usr/local/share/nvim/runtime/ftplugin.vim
004.647  000.197  000.197: sourcing /usr/local/share/nvim/runtime/indent.vim
004.654  000.748: sourcing vimrc file(s)
004.933  000.261  000.261: sourcing /usr/local/share/nvim/runtime/filetype.lua
005.397  000.222  000.222: sourcing /usr/local/share/nvim/runtime/syntax/synload.vim
005.470  000.511  000.289: sourcing /usr/local/share/nvim/runtime/syntax/syntax.vim
006.275  000.337  000.337: sourcing /usr/local/share/nvim/runtime/plugin/gzip.vim
007.153  000.322  000.322: sourcing /usr/local/share/nvim/runtime/pack/dist/opt/matchit/plugin/matchit.vim
007.184  000.892  000.570: sourcing /usr/local/share/nvim/runtime/plugin/matchit.vim
007.540  000.333  000.333: sourcing /usr/local/share/nvim/runtime/plugin/matchparen.vim
007.988  000.430  000.430: sourcing /usr/local/share/nvim/runtime/plugin/netrwPlugin.vim
008.219  000.205  000.205: sourcing /usr/local/share/nvim/runtime/plugin/rplugin.vim
008.440  000.198  000.198: sourcing /usr/local/share/nvim/runtime/plugin/shada.vim
008.650  000.185  000.185: sourcing /usr/local/share/nvim/runtime/plugin/spellfile.vim
008.899  000.228  000.228: sourcing /usr/local/share/nvim/runtime/plugin/tarPlugin.vim
009.104  000.178  000.178: sourcing /usr/local/share/nvim/runtime/plugin/tutor.vim
009.383  000.254  000.254: sourcing /usr/local/share/nvim/runtime/plugin/zipPlugin.vim
009.601  000.193  000.193: sourcing /usr/local/share/nvim/runtime/plugin/editorconfig.lua
009.836  000.213  000.213: sourcing /usr/local/share/nvim/runtime/plugin/man.lua
010.065  000.208  000.208: sourcing /usr/local/share/nvim/runtime/plugin/osc52.lua
010.291  000.206  000.206: sourcing /usr/local/share/nvim/runtime/plugin/tohtml.lua
010.305  000.818: loading rtp plugins
010.372  000.067: loading packages
010.515  000.143: loading after plugins
010.522  000.006: inits 3
010.523  000.001: reading ShaDa
010.589  000.066: opening buffers
010.602  000.014: BufEnter autocommands
010.604  000.002: editing files in windows
010.670  000.066: VimEnter autocommands
010.672  000.002: UIEnter autocommands
010.673  000.001: before starting main loop
010.841  000.168: first screen update
010.842  000.002: --- NVIM STARTED ---

```

Without ripgrep installed:

```
--- Startup times for process: Primary (or UI client) ---

times in msec
 clock   self+sourced   self:  sourced script
 clock   elapsed:              other lines

000.000  000.000: --- NVIM STARTING ---
000.095  000.094: event init
000.163  000.068: early init
000.294  000.131: locale set
000.330  000.036: init first window
000.613  000.283: inits 1
000.618  000.006: window checked
--- Startup times for process: Primary (or UI client) ---

times in msec
 clock   self+sourced   self:  sourced script
 clock   elapsed:              other lines

000.000  000.000: --- NVIM STARTING ---
000.112  000.112: event init
000.207  000.095: early init
000.336  000.129: locale set
000.374  000.038: init first window
000.646  000.272: inits 1
000.651  000.005: window checked
000.654  000.003: parsing arguments
000.974  000.025  000.025: require('vim.shared')
001.037  000.024  000.024: require('vim.inspect')
001.076  000.034  000.034: require('vim._options')
001.078  000.103  000.044: require('vim._editor')
001.079  000.144  000.017: require('vim._init_packages')
001.080  000.281: init lua interpreter
002.518  001.438: nvim_ui_attach
002.764  000.247: nvim_set_client_info
002.766  000.001: --- NVIM STARTED ---

--- Startup times for process: Embedded ---

times in msec
 clock   self+sourced   self:  sourced script
 clock   elapsed:              other lines

000.000  000.000: --- NVIM STARTING ---
000.078  000.077: event init
000.146  000.068: early init
000.302  000.156: locale set
000.334  000.032: init first window
000.619  000.285: inits 1
000.627  000.008: window checked
000.630  000.004: parsing arguments
000.927  000.025  000.025: require('vim.shared')
000.989  000.023  000.023: require('vim.inspect')
001.019  000.025  000.025: require('vim._options')
001.021  000.091  000.043: require('vim._editor')
001.021  000.147  000.030: require('vim._init_packages')
001.022  000.246: init lua interpreter
001.057  000.034: expanding arguments
001.068  000.012: inits 2
001.276  000.207: init highlight
001.278  000.002: waiting for UI
001.367  000.089: done waiting for UI
001.375  000.008: clear screen
001.464  000.009  000.009: require('vim.keymap')
042.800  041.423  041.413: require('vim._defaults')
042.807  000.009: init default mappings & autocommands
043.129  000.072  000.072: sourcing /usr/local/share/nvim/runtime/ftplugin.vim
043.166  000.017  000.017: sourcing /usr/local/share/nvim/runtime/indent.vim
043.189  000.292: sourcing vimrc file(s)
043.357  000.149  000.149: sourcing /usr/local/share/nvim/runtime/filetype.lua
043.477  000.056  000.056: sourcing /usr/local/share/nvim/runtime/syntax/synload.vim
043.553  000.164  000.108: sourcing /usr/local/share/nvim/runtime/syntax/syntax.vim
044.316  000.475  000.475: sourcing /usr/local/share/nvim/runtime/plugin/gzip.vim
045.083  000.366  000.366: sourcing /usr/local/share/nvim/runtime/pack/dist/opt/matchit/plugin/matchit.vim
045.122  000.788  000.422: sourcing /usr/local/share/nvim/runtime/plugin/matchit.vim
045.263  000.123  000.123: sourcing /usr/local/share/nvim/runtime/plugin/matchparen.vim
045.734  000.453  000.453: sourcing /usr/local/share/nvim/runtime/plugin/netrwPlugin.vim
045.974  000.213  000.213: sourcing /usr/local/share/nvim/runtime/plugin/rplugin.vim
046.032  000.038  000.038: sourcing /usr/local/share/nvim/runtime/plugin/shada.vim
046.066  000.012  000.012: sourcing /usr/local/share/nvim/runtime/plugin/spellfile.vim
046.332  000.248  000.248: sourcing /usr/local/share/nvim/runtime/plugin/tarPlugin.vim
046.546  000.184  000.184: sourcing /usr/local/share/nvim/runtime/plugin/tutor.vim
046.846  000.276  000.276: sourcing /usr/local/share/nvim/runtime/plugin/zipPlugin.vim
046.981  000.094  000.094: sourcing /usr/local/share/nvim/runtime/plugin/editorconfig.lua
047.137  000.136  000.136: sourcing /usr/local/share/nvim/runtime/plugin/man.lua
047.300  000.144  000.144: sourcing /usr/local/share/nvim/runtime/plugin/osc52.lua
047.389  000.061  000.061: sourcing /usr/local/share/nvim/runtime/plugin/tohtml.lua
047.406  000.658: loading rtp plugins
047.488  000.083: loading packages
047.671  000.183: loading after plugins
047.680  000.009: inits 3
047.686  000.006: reading ShaDa
047.785  000.099: opening buffers
047.804  000.019: BufEnter autocommands
047.805  000.002: editing files in windows
047.896  000.091: VimEnter autocommands
047.899  000.002: UIEnter autocommands
047.900  000.001: before starting main loop
048.146  000.247: first screen update
048.148  000.002: --- NVIM STARTED ---
```

So it's 40ms extra from `_defaults` in my PC. This only happens on WSL, not in native windows.

My guess is that if WSL doesn't find the binary in WSL it starts looking at Windows (there are many `/mnt/c/...` directories in WSL path), and it's widely known that using/executing windows files in WSL is slow (that's why using `vim.ui.open()` is also slow on WSL if it has to use explorer.exe or the Windows browser).

**Comment by lewis6991** [comment]
Is there any way to make wsl not do that?

**Comment by LunarLambda** [comment]
If we have a means to check for `if wsl` we could just restrict the search to `$CARGO_HOME/bin`, `/usr/bin` and `/usr/local/bin`? We probably don't want to touch WSL's `$PATH`, so I'm not sure how much we can do to avoid this. But admittedly I'm not an expert on WSL matters.

**Comment by clason** [comment]
Gotta ask: Is simply installing ripgrep not an option? You'll get (much) faster grep _and_ startup -- what's not to like?

**Comment by lewis6991** [comment]
This is a general problem with `executable()` which we use in other places so this issue works well to track that.

**Comment by kyoh86** [comment]
If you want to identify a WSL, you can do so by the presence of several environment variables.
- WSL_DISTRO_NAME
- WSLENV
- etc.
Perhaps `WSL_DISTRO_NAME` is appropriate to detect that is in WSL.

**Comment by SebasF1349** [comment]
Idk what the solution should be, but I'm not sure the best solution is to make `executable()` only look for executables on WSL and not on Windows. Yes, it would make it faster, but it will mean WSL users to lose features. For example, now users can use `gx` to open a directory in explorer.exe. Or, even better, to open a link in my browser. If neovim restricts where `executable()` looks for, those features will be lost and users will have to install a new browser on WSL just for that (which in my own case it doesn't mean just install a new browser, but also maintain configuration - like extensions, keymaps, open tabs - between both).

**Comment by lewis6991** [comment]
A valid solution here is to do nothing if it's deemed that not searching in PATH is not acceptable.

**Comment by LunarLambda** [comment]
Out of curiousity, this only applies to a Linux build of Neovim being run inside of WSL, right?
Is there a strong argument against using a Windows build of Neovim in WSL instead? You can mount/access WSL's filesystem via Windows filesystem APIs, and tools like `git`, `curl`, and various LSPs should be available natively too.

While not the same thing as WSL obviously, I used to run Neovim for Windows inside of an MSYS2 install, and it worked pretty well for the most part.

**Comment by SebasF1349** [comment]
If the problem is accessing programs (like rg) in windows, idk how it will be better accessing another program (neovim) in windows. WSL is just slow when having to do stuff outside of WSL.

**Comment by dundargoc** [comment]
> For example, now users can use gx to open a directory in explorer.exe. Or, even better, to open a link in my browser.

But that is a special case, because browsers are more useful outside of wsl. For any other program like git or ripgrep this isn't really true.

**Comment by kuuote** [comment]
I think fundamentally problem is not way to disable this behavior in user configuration.
Currently defaults lua is loaded before init.vim or init.lua.

**Comment by Shougo** [comment]
The performance problem exists even in Linux environment.
It is about some millisecond, but it is not acceptable overhead for me.

NOTE: My neovim startup time is 9ms.

But I think the real problem is that `defaults.lua` is executed before `init.vim` or `init.lua` and users cannot disable it.
I even know users who are patching defaults.lua to disable it. Isn't this counterproductive?

If defaults.lua checks `grepprg` value and it is not set by user, the set is OK.

**Comment by zeertzjq** [comment]
Also cc @gpanders

**Comment by gpanders** [comment]
One possible short term course of action would be to move the default grepprg setting out of _defaults.lua and into a plugin file, which can be opted out of somehow.

**Comment by clason** [comment]
Note that we'll run into similar problems with other (harder) external requirements like `curl` and `tar`, so the bigger issue worth tracking here is not limited to `_defaults.lua`.

**Comment by mfussenegger** [comment]
Another potential solution could be to resolve the default lazy on first use. No idea what the current representation looks like but could do something like in https://github.com/neovim/neovim/pull/25123 if we currently don't know if the user changed the default

**Comment by LunarLambda** [comment]
I'll just say I would be happy to implement any solution deemed acceptable as far as _defaults.lua is concerned ^^

**Comment by Shougo** [comment]
Not limited to this option, performing a high-cost operation like `executable()` in `_defaults.lua`, which is always executed before user settings, should be avoided.

Regarding this process, I think the only option is to simply remove the check. However, if such a process becomes necessary in other functions, I believe the only solution is to delay it.

**Comment by dundargoc** [comment]
Unsure why this discussion is so defaults.lua-centric, this problem already existed before it and is infinitely worse if user doesn't have a clipboard provider as a bunch of random providers will be searched with executable() calls, one for each provider.

**Comment by Shougo** [comment]
> Unsure why this discussion is so defaults.lua-centric, this problem already existed before it and is infinitely worse if user doesn't have a clipboard provider as a bunch of random providers will be searched with `executable()` calls, one for each provider.

Oh, it is another problem.

I disable clipboard support.  Because it is slow.

```vim
" Disable clipboard support
set clipboard=
```

But this search only executed when clipboard is used?  It seems executed lazily.

**Comment by justinmk** [comment]
Let's add a parameter to `executable()` which makes it fast. https://github.com/fish-shell/fish-shell/commit/3374692b9113e85c690059ac3def5f4aa190294f shows a strategy:

> If the cmd we are executing does
not contain a period (i.e. has no extension) it by definition cannot be a
Windows executable. In this case, we skip searching for it in any of the
auto-mounted, auto-PATH-appended directories like `/mnt/c/Windows/` or
`/mnt/c/Program Files`, but we *do* include those directories if what we're
searching for could be a Windows executable.

That might not even require a flag, just improving the `executable()` logic.

**Comment by LunarLambda** [comment]
How does this interact with the PATHEXT variable on Windows (a list of file extensions the system tries when looking for a command `foo`, e.g. it tries `foo.bat`, `foo.cmd`, `foo.exe`, etc..)

Does WSL not do this? Or would windows/wsl user code need to explicitly specify an extension going forward?

**Comment by kyoh86** [comment]
I understand that the essence of the problem is the slowness of executable(), but the starting point of this Issue is that the decision process in defaults.lua is taking a long time.
If it is difficult to resolve the essence of the issue, could you please take measures to work around the problem that started it in the first place?
If you understand that the process is too heavy to be forced to call in defaults.lua, it is not unreasonable to conclude that it should not be in defaults.lua.

**Comment by justinmk** [comment]
I already suggested an approach, above.

**Comment by Shougo** [comment]
Yes.  I know.  It seems the right approach.  But it is not implemented yet.
And it is for WSL environment fix.  It is not for Windows environment.

I think workaround fix is needed for the problem.

**Comment by LunarLambda** [comment]
Here's an idea: Can we make `grepprg` have the ability to be set to an expression? Similar to `statusline`, etc?

Then lua code could lazily evaluate which program to use based on availability on the first `:grep` invocation, and cache the result. Zero cost to startuptime and zero cost to the user if they don't use the command.

**Comment by lewis6991** [comment]
`keywordprg` can be prefixed with `:` to be interpreted as an ex command. We could do the same with `grepprg` as a short term solution.
